contrib/connector/fc/fibreChannel.go

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2018 Huawei Technologies Co., Ltd. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fc
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"strconv"
+	"strings"
+	"time"
+)
+
+type fibreChannel struct {
+	helper *linuxfc
+}
+
+func (f *fibreChannel) connectVolume(conn map[string]interface{}) (map[string]string, error) {
+	hbas, err := f.getFChbasInfo()
+	if err != nil {
+		return nil, err
+	}
+	volPaths := f.getVolumePaths(conn, hbas)
+	if len(volPaths) == 0 {
+		errMsg := fmt.Sprintf("No fc devices found.")
+		log.Println(errMsg)
+		return nil, errors.New(errMsg)
+	}
+
+	var tries = 3
+	devicePath, deviceName := f.volPathDiscovery(volPaths, tries, conn["target_wwn"].([]string), hbas)
+	if devicePath != "" && deviceName != "" {
+		log.Printf("Found fibre channel volume name, devicePath is %s, deviceName is %s", devicePath, deviceName)
+	}
+
+	deviceWWN, err := f.helper.getSCSIWWN(devicePath)
+	if err != nil {
+		return nil, err
+	}
+
+	return map[string]string{"scsi_wwn": deviceWWN, "path": devicePath}, nil
```

Comment:
There is no support for multipath?

Reply:
I will add multi-path and multi-attach at the same time later. Here i just implemented the basic functions.